File: comguanmuonehours/src/main/java/com/guanmu/onehours/model/MTag.java
package com.guanmu.onehours.model;

import com.guanmu.onehours.DLinkTagPoint;
import com.guanmu.onehours.DLinkTagPointDao;
import com.guanmu.onehours.DTag;
import com.guanmu.onehours.DTagDao;
import com.guanmu.onehours.DaoSession;

import org.greenrobot.greendao.query.QueryBuilder;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.UUID;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

/**
 * Created by Larry.
 */
public class MTag {
    private DaoSession daoSession;
    private DTagDao dao;
    private DLinkTagPointDao pointDao;

    private String name;
    private String belong;

    private MTag parent;
    private List<MTag> children;
    private List<MPoint> points;

    public MTag(DaoSession dao) {
        this(dao, null, null, null);
    }

    public MTag(DaoSession dao, String name, String belong) {
        this(dao,
                tryGetGUID(dao.getDTagDao(), name, belong),
                name,
                belong);
    }

    private MTag(DaoSession dao, DTag dTag) {
        this(dao,
                UUID.fromString(dTag.getGUID()),
                dTag.getName(),
                dTag.getBelong()
        );
    }

    private MTag(DaoSession dao, UUID GUID, String name, String belong) {
        this.daoSession = dao;
        this.dao = dao.getDTagDao();
        this.pointDao = dao.getDLinkTagPointDao();
        this.name = name != null ? name : "";
        this.belong = belong;

        children = new ArrayList<>();
        points = new ArrayList<>();
        save();
    }


    public static List<MTag> loadAll(DaoSession dao) {
        return loadAll(dao, false, null);
    }

    public static List<MTag> loadAll(DaoSession dao, boolean filterBelong, @Nullable String belong) {
        cleanUnrelated(dao);
        QueryBuilder<DTag> qb = dao.getDTagDao().queryBuilder();
        if (filterBelong) {
            if (belong == null) {
                qb = qb.where(DTagDao.Properties.Belong.isNull());
            } else {
                qb = qb.where(DTagDao.Properties.Belong.eq(belong));
            }
        }
        return listFrom(dao, qb.list());
    }

    private static void cleanUnrelated(DaoSession dao) {
        for (DTag dTag
                : dao.getDTagDao().loadAll()) {
            if (dTag.getDLinkPointList().isEmpty()) {
                dTag.delete();
            }
        }
    }


    private static UUID tryGetGUID(DTagDao dao, String name, String belong) {
        try {
            QueryBuilder<DTag> qb = dao.queryBuilder().where(DTagDao.Properties.Name.eq(name));
            if (belong == null) {
                qb = qb.where(DTagDao.Properties.Belong.isNull());
            } else {
                qb = qb.where(DTagDao.Properties.Belong.eq(belong));
            }
            return UUID.fromString(qb.unique().getGUID());
        } catch (NullPointerException ignored) {
            return null;
        }
    }

    static List<MTag> listFrom(DaoSession daoSession, @Nullable List<DTag> dTagList) {
        List<MTag> result = new ArrayList<>();
        if (dTagList != null) {
            for (DTag dTag :
                    dTagList) {
                result.add(new MTag(daoSession, dTag));
            }
        }
        return result;
    }

    public UUID getGUID() {
        return UUID.fromString(getD().getGUID());
    }

    @NonNull
    private DTag getD() {
        UUID GUID = tryGetGUID(dao, name, belong);
        if (GUID != null) {
            return dao.load(GUID.toString());
        } else {
            DTag dTag = new DTag(UUID.randomUUID().toString());
            dTag.setName(name);
            dTag.setBelong(belong);
            if (parent != null) {
                dTag.setParent(parent.getD());
            }
            dao.insert(dTag);
            return dTag;
        }
    }


    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getBelong() {
        return belong;
    }

    public void setBelong(String belong) {
        this.belong = belong;
    }

    public List<MPoint> getPoints() {
        return MPoint.listFrom(daoSession, getD().getDPointList());
    }

    public void setPoint(@NonNull MPoint... points) {
        setPoint(Arrays.asList(points));
    }

    public void setPoint(@NonNull List<MPoint> points) {
        removePoint(MPoint.listFrom(daoSession, getD().getDPointList()));
        addPoint(points);
    }

    public void addPoint(@NonNull MPoint... points) {
        addPoint(Arrays.asList(points));
    }

    public void addPoint(@NonNull List<MPoint> points) {
        DTag dTag = getD();
        QueryBuilder qb = pointDao.queryBuilder().where(DLinkTagPointDao.Properties.TagId.eq(dTag.getGUID()));
        for (MPoint point :
                points) {
            if (qb.where(DLinkTagPointDao.Properties.PointId.eq(point.getGUID().toString())).count() == 0) {
                DLinkTagPoint link = new DLinkTagPoint();
                link.setPointId(point.getGUID().toString());
                link.setTagId(dTag.getGUID());
                pointDao.save(link);
            }
        }
    }

    public void removePoint(@NonNull MPoint... points) {
        removePoint(Arrays.asList(points));
    }

    public void removePoint(@NonNull List<MPoint> points) {
        List<String> pointsId = new ArrayList<>();
        for (MPoint point :
                points) {
            pointsId.add(point.getGUID().toString());
        }
        pointDao.queryBuilder()
                .where(DLinkTagPointDao.Properties.TagId.eq(getD().getGUID()))
                .where(DLinkTagPointDao.Properties.PointId.in(pointsId))
                .buildDelete().executeDeleteWithoutDetachingEntities();
    }

    public List<MTag> getChildren() {
        return listFrom(daoSession, getD().getChildren());
    }

    public void setChildren(@NonNull List<MTag> children) {
        removeChildren(listFrom(daoSession, getD().getChildren()));
        addChildren(children);
    }

    public void setChildren(@NonNull MTag... children) {
        setChildren(Arrays.asList(children));
    }

    public void addChildren(@NonNull MTag... children) {
        addChildren(Arrays.asList(children));
    }

    public void addChildren(@NonNull List<MTag> children) {
        for (MTag child :
                children) {
            child.setParent(this);
        }
    }

    public void removeChildren(@NonNull MTag... children) {
        removeChildren(Arrays.asList(children));
    }

    public void removeChildren(@NonNull List<MTag> children) {
        for (MTag child :
                children) {
            if (child.getParent() == this) {
                child.setParent(null);
            }
        }
    }

    public MTag getParent() {
        return new MTag(daoSession, getD().getParent());
    }

    public void setParent(@Nullable MTag parent) {
        DTag dTag = getD();
        if (parent != null) {
            dTag.setParentId(parent.getD().getGUID());
        }
    }

    public void save() {
        DTag dTag = getD();
        dTag.setName(name);
        dTag.setBelong(belong);
        dao.save(dTag);
    }

    public void delete() {
        UUID GUID = tryGetGUID(dao, name, belong);
        if (GUID != null) {
            dao.deleteByKey(GUID.toString());
        }
    }

    @Override
    public String toString() {
        return name;
    }
}

File: comguanmuonehours/src/main/java/com/guanmu/onehours/model/MTag.kt
package com.guanmu.onehours.model

import com.guanmu.onehours.*
import java.util.*

/**
 * Created by Larry.
 */
class MTag constructor(private val daoSession: DaoSession, name: String?, var belong: String?) {
    private val dao: DTagDao = daoSession.dTagDao
    private val pointDao: DLinkTagPointDao = daoSession.dLinkTagPointDao

    var name: String = ""

    private val parent: MTag? = null
    private val children: List<MTag>
    private val points: List<MPoint>

    val guid: UUID
        get() = UUID.fromString(d.guid)

    private val d: DTag
        get() {
            val GUID = tryGetGUID(dao, name, belong)
            return if (GUID != null) {
                dao.load(GUID.toString())
            } else {
                val dTag = DTag(UUID.randomUUID().toString())
                dTag.name = name
                dTag.belong = belong
                if (parent != null) {
                    dTag.parent = parent.d
                }
                dao.insert(dTag)
                dTag
            }
        }

    private constructor(dao: DaoSession, dTag: DTag) : this(dao,
            dTag.name,
            dTag.belong
    )

    init {
        this.name = name ?: ""

        children = ArrayList()
        points = ArrayList()
        save()
    }

    fun getPoints(): List<MPoint> {
        return MPoint.listFrom(daoSession, d.dPointList)
    }

    fun setPoint(vararg points: MPoint) {
        setPoint(listOf(*points))
    }

    fun setPoint(points: List<MPoint>) {
        removePoint(MPoint.listFrom(daoSession, d.dPointList))
        addPoint(points)
    }

    fun addPoint(vararg points: MPoint) {
        addPoint(listOf(*points))
    }

    fun addPoint(points: List<MPoint>) {
        val dTag = d
        val qb = pointDao.queryBuilder().where(DLinkTagPointDao.Properties.TagId.eq(dTag.guid))
        for (point in points) {
            if (qb.where(DLinkTagPointDao.Properties.PointId.eq(point.guid!!.toString())).count() == 0L) {
                val link = DLinkTagPoint()
                link.pointId = point.guid!!.toString()
                link.tagId = dTag.guid
                pointDao.save(link)
            }
        }
    }

    fun removePoint(vararg points: MPoint) {
        removePoint(listOf(*points))
    }

    fun removePoint(points: List<MPoint>) {
        val pointsId = ArrayList<String>()
        for (point in points) {
            pointsId.add(point.guid!!.toString())
        }
        pointDao.queryBuilder()
                .where(DLinkTagPointDao.Properties.TagId.eq(d.guid))
                .where(DLinkTagPointDao.Properties.PointId.`in`(pointsId))
                .buildDelete().executeDeleteWithoutDetachingEntities()
    }

    fun getChildren(): List<MTag> {
        return listFrom(daoSession, d.children)
    }

    fun setChildren(children: List<MTag>) {
        removeChildren(listFrom(daoSession, d.children))
        addChildren(children)
    }

    fun setChildren(vararg children: MTag) {
        setChildren(listOf(*children))
    }

    fun addChildren(vararg children: MTag) {
        addChildren(listOf(*children))
    }

    fun addChildren(children: List<MTag>) {
        for (child in children) {
            child.setParent(this)
        }
    }

    fun removeChildren(vararg children: MTag) {
        removeChildren(listOf(*children))
    }

    fun removeChildren(children: List<MTag>) {
        for (child in children) {
            if (child.getParent() === this) {
                child.setParent(null)
            }
        }
    }

    fun getParent(): MTag {
        return MTag(daoSession, d.parent)
    }

    fun setParent(parent: MTag?) {
        val dTag = d
        if (parent != null) {
            dTag.parentId = parent.d.guid
        }
    }

    fun save() {
        val dTag = d
        dTag.name = name
        dTag.belong = belong
        dao.save(dTag)
    }

    fun delete() {
        val GUID = tryGetGUID(dao, name, belong)
        if (GUID != null) {
            dao.deleteByKey(GUID.toString())
        }
    }

    override fun toString(): String {
        return name
    }

    companion object {

        @JvmOverloads
        fun loadAll(dao: DaoSession, filterBelong: Boolean = false, belong: String? = null): List<MTag> {
            cleanUnrelated(dao)
            var qb = dao.dTagDao.queryBuilder()
            if (filterBelong) {
                qb = qb.where(if (belong == null) {
                    DTagDao.Properties.Belong.isNull
                } else {
                    DTagDao.Properties.Belong.eq(belong)
                })
            }
            return listFrom(dao, qb.list())
        }

        private fun cleanUnrelated(dao: DaoSession) {
            for (dTag in dao.dTagDao.loadAll()) {
                if (dTag.dLinkPointList.isEmpty()) {
                    dTag.delete()
                }
            }
        }


        private fun tryGetGUID(dao: DTagDao, name: String?, belong: String?): UUID? {
            return try {
                var qb = dao.queryBuilder().where(DTagDao.Properties.Name.eq(name))
                qb = qb.where(if (belong == null) {
                    DTagDao.Properties.Belong.isNull
                } else {
                    DTagDao.Properties.Belong.eq(belong)
                })
                UUID.fromString(qb.unique().guid)
            } catch (ignored: NullPointerException) {
                null
            }

        }

        internal fun listFrom(daoSession: DaoSession, dTagList: List<DTag>?): List<MTag> {
            val result = ArrayList<MTag>()
            if (dTagList != null) {
                for (dTag in dTagList) {
                    result.add(MTag(daoSession, dTag))
                }
            }
            return result
        }
    }
}
